Mathlib/LinearAlgebra/Reflection.lean

```diff
@@ -114,6 +121,173 @@ lemma bijOn_reflection_of_mapsTo {Φ : Set M} (h : f x = 2) (h' : MapsTo (reflec
+lemma reflection_mul_reflection_pow_apply (m : ℕ) (z : M) :
+    ((reflection hf * reflection hg) ^ m) z =
+      z +
+        (S_eval_t ((m - 2) / 2) * (S_eval_t ((m - 1) / 2) + S_eval_t ((m - 3) / 2))) •
+          ((g x * f z - g z) • y - f z • x) +
+        (S_eval_t ((m - 1) / 2) * (S_eval_t (m / 2) + S_eval_t ((m - 2) / 2))) •
+          ((f y * g z - f z) • x - g z • y) := by
+  induction m with
+  | zero => simp
+  | succ m ih =>
+    /- Now, let us collect two facts about the sequence `S_eval_t`. These easily follow from the
+    properties of the `S` polynomials. -/
+    have S_eval_t_sub_two (k : ℤ) : S_eval_t (k - 2) = t * S_eval_t (k - 1) - S_eval_t k := by
+      simp [S_sub_two]
+    have S_eval_t_sq_add_S_eval_t_sq (k : ℤ) :
+        S_eval_t k ^ 2 + S_eval_t (k + 1) ^ 2 - t * S_eval_t k * S_eval_t (k + 1) = 1 := by
+      simpa using congr_arg (Polynomial.eval t) (S_sq_add_S_sq R k)
+    -- Apply the inductive hypothesis.
+    rw [pow_succ', LinearEquiv.mul_apply, ih, LinearEquiv.mul_apply]
+    -- Expand out all the reflections and use `hf`, `hg`.
+    simp only [reflection_apply, map_add, map_sub, map_smul, hf, hg]
+    -- `m` can either be written in the form `2 * k` or `2 * k + 1`.
+    obtain ⟨k, even | odd⟩ := Int.even_or_odd' m
+    · /- The goal contains many expressions of the form `S_eval_t n`. We may rewrite all of them as
+      `S_eval_t (-2 + k)`, `S_eval_t (-1 + k)`, or `S_eval_t k`. -/
+      push_cast
+      simp_rw [even, add_sub_assoc (2 * k), sub_eq_add_neg (2 * k), add_comm (2 * k),
+        add_mul_ediv_left _ k (by norm_num : (2 : ℤ) ≠ 0)]
+      norm_num
+      /- Now, equate the coefficients on both sides. These linear combinations were
+      found using `polyrith`. -/
+      match_scalars
+      · rfl
+      · linear_combination (norm := ring_nf) (-g z * f y * S_eval_t (-1 + k) +
+            f z * S_eval_t (-1 + k)) * S_eval_t_sub_two k +
+          (-g z * f y + f z) * S_eval_t_sq_add_S_eval_t_sq (k - 1)
+      · linear_combination (norm := ring_nf) g z * S_eval_t (-1 + k) * S_eval_t_sub_two k +
+          g z * S_eval_t_sq_add_S_eval_t_sq (k - 1)
+    · /- The goal contains many expressions of the form `S_eval_t n`. We may rewrite all of them as
+      `S_eval_t (-1 + k)`, `S_eval_t k`, or `S_eval_t (1 + k)`. -/
+      push_cast
+      simp_rw [odd, add_assoc (2 * k), add_sub_assoc (2 * k), add_comm (2 * k),
+        add_mul_ediv_left _ k (by norm_num : (2 : ℤ) ≠ 0)]
+      norm_num
+      /- Now, equate the coefficients on both sides. These linear combinations were
+      found using `polyrith`. -/
+      match_scalars
+      · rfl
+      · linear_combination (norm := ring_nf) (-g z * f y * S_eval_t k +
+            f z * S_eval_t k) * S_eval_t_sub_two (k + 1) +
+          (-g z * f y + f z) * S_eval_t_sq_add_S_eval_t_sq (k - 1)
+      · linear_combination (norm := ring_nf) g z * S_eval_t k * S_eval_t_sub_two (k + 1) +
+          g z * S_eval_t_sq_add_S_eval_t_sq (k - 1)
+
```

Comment:
It might be useful to have an unapplied version of the lemma:
```
/-- A formula for $(r_1 r_2)^m$, where $m$ is a natural number. -/
lemma reflection_mul_reflection_pow' (m : ℕ) (t : R) (ht : t = f y * g x - 2) :
    ((reflection hf * reflection hg) ^ m) =
      LinearMap.id (R := R) (M := M) +
        (((S R ((m - 2) / 2) * (S R ((m - 1) / 2) + S R ((m - 3) / 2))).eval t) •
          ((g x • f - g).smulRight y - f.smulRight x)) +
        (((S R ((m - 1) / 2) * (S R (m / 2) + S R ((m - 2) / 2))).eval t) •
          ((f y • g - f).smulRight x - g.smulRight y)) := by
  ext z
  simp
  rw [← reflection_mul_reflection_pow_apply hf hg m z t ht]
  rw [LinearMap.pow_apply, LinearEquiv.pow_apply]
  rfl
```